### DISCUSSION TITLE
Starlark Transition Composition

### DISCUSSION THREAD

**Discussion by katre** [bazelbuild/bazel#22019] [Design proposals]
We've had native transition composition for a long time, let's make that available to Starlark users!

Design proposal here: https://github.com/bazelbuild/proposals/blob/main/designs/2024-04-16-transition-composition.md

Tagging @gregestren as reviewer, but anyone with comments feel free to contribute.

**Comment by fmeum** [comment]
I am overall very supportive of this addition, it would help clean up a bunch of indirections required to combine an exec transition with a non-exec transition.

> 3. The actual configuration change applies the first transition, and then applies the second transition to each split of the result.

What if the second transition in `transition1 + transition2` is the split transition? It could make sense to formalize this section with references to the individual transitions in such a composition.

> 2. The string "exec", for the execution transition on the default exec group.

Now that there is `config.exec`, I'm not a fan of this kind of syntactic sugar. `transition + "exec"` looks magical (or even wrong) to me and breaks in a very bad way as soon as `transition == "target"`. It also requires special-casing in LSPs.

**Reply by katre** [reply]
>  It could make sense to formalize this section with references to the individual transitions in such a composition.

Makes sense, I'll beef this up shortly.

**Reply by katre** [reply]
>  I'm not a fan of this kind of syntactic sugar. `transition + "exec"` looks magical

I agree with this, but given that we currently accept "exec" and "target" for `cfg`, I think it's better to keep the usages matching. It will also ease refactorings, I think, to change like this:

```
- "_some_tool": attr.label(cfg = "exec"),
+ "_some_tool": attr.label(cfg = custom_trans + "exec"),
```

I **don't** think we have expressiveness in Starlark to only allow `transition + "exec"` in an attribute directly, so we need to also support it in all contexts.

Based on my reading of [`EvalUtils.binaryOp`](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/net/starlark/java/eval/EvalUtils.java;drc=16b82af1b7eac1a13fa73390a68fa03297a29630;l=378), we could actually support `"exec" + custom_transition`, but I think that usage is a terrible idea and we shouldn't allow it that way.

**Reply by fmeum** [reply]
Why is `transition + "exec"` okay but not `"exec" + transition`? While I don't expect `+` to be commutative, I wouldn't expect a type error if I flip the order.

**Reply by katre** [reply]
So, "terrible idea" is a bit strong, probably. I think "string + transition" is more confusing than "transition + string". Possibly a better choice of operator would make this easier to think about.

**Reply by gregestren** [reply]
Why does `transition == "target"` break more than any other transition?

**Reply by gregestren** [reply]
I also find it user-unintuitive why `transition + "exec"` works but `"exec" + transition` doesn't.  What would the APi documentation look like to make that intuitively clear?

I'm a bit worried combining `exec` with anything is going to be confusing, since `exec` is "special". For example, `exec` [inherits option defaults](https://github.com/bazelbuild/bazel/blob/76a214932da4a9dcf73d3cfe9313e0172d20e630/src/main/java/com/google/devtools/build/lib/analysis/starlark/FunctionTransitionUtil.java#L146-L165), not the previous options. Someone could write `my_trans = trans1 + "exec"` where `trans1` sets `--foo=bar` then `exec` implicitly empties it, so the final output is `--foo=""`. Is that what they expected? 

I'm also unsure how cleanly Bazel's exec implementation will integrate. Since the implementation has special logic to determine if a config is the exec config, a composed transition that includes the exec config might or might not qualify. I'm not sure, for example, how this affects output path naming: I'd think any composed exec transition would not be considered the base exec transition, so its path would have `ST-<hash>`. That sounds okay but just a thing to check.

More fundamentally, it's possible this composition will totally break without some technical adjustments. I think there's a good chance it'll work. But it's definitely something we'd want to test thoroughly.

**Reply by gregestren** [reply]
> Someone could write my_trans = trans1 + "exec" where trans1 sets --foo=bar then exec implicitly empties it, so the final output is --foo=""

To clarify: if someone wants to apply the exec transition and also set Starlark flag `--//foo=1`, this hypothetical syntax would work:

`"exec" + trans`

but this wouldn't

`trans + "exec"`

We could conceivably modify those semantics. But without special new logic taking into account composition that's the behavior we'd see.

**Reply by katre** [reply]
To clarify further: the semantics of the exec transition are that `trans + "exec"` would effectively only be the exec transition, because the current exec transition resets almost all flags to stable values.

**Reply by fmeum** [reply]
> Why does transition == "target" break more than any other transition?

If you have some code in Starlark that does `transition + "exec"` for a `transition` obtained from elsewhere and someone passes in `"target"` or `"exec"`, the result will suddenly be a string such as `"execexec"`, not a transition.

**Reply by gregestren** [reply]
@fmeum which indirections specifically inspired you?

IMO 99% of the conceptual complexity in this proposal is the exec transition integration. We could even do wild things like invoke a build with a custom exec transition that's defined as a composition of `trans1` plus the default exec transition. I think these issues scope-bleed into wider philosophy about what we want the exec transition to be.

If the use cases you're thinking about aren't trying to actually "modify" the exec transition but remain clear cases of "apply arbitrary transition B, then apply the exec transition, recognizing these are unrelated transitions and just happen to be sequenced together", that's IMO easier to reason about now.

**Reply by katre** [reply]
@fmeum You're right , `"exec" + "target"` would give the wrong result, which is unfortunate.

To clarify, does this mean you support:
1. Don't allow using shorthand names `"exec"` and `"target"` with composition (and expose something like `config.target` to mean the no-op transition)?
2. Support both `transition + "exec"` and `"exec" + transition` (but not, sadly, `"exec" + "exec"`)?

**Reply by fmeum** [reply]
I support 1. I don't know whether we even need the `config.target` noop, I would consider usages such as `transition + config.exec() if transition else config.exec()` reasonable as well.

**Comment by sluongng** [comment]
Nit: I don't like using the plus `+` operator here as this does not have the commutative property that's often a part of addition.

Could we use a different operator? If it's not allowed in Starlark, I'm ok with a new built-in function as well.

**Reply by fmeum** [reply]
> Nit: I don't like using the plus + operator here as this does not have the commutative property that's often a part of addition.

I see a point for `+` being too magical here, but keep in mind that `+` isn't commutative for strings and lists either. In fact, even floating point additions may not be commutative.

**Reply by fmeum** [reply]
`|` could be an alternative.

**Reply by sluongng** [reply]
> keep in mind that + isn't commutative for strings and lists either.

Ah! True that. 

However, thinking of how the second operand could be either a special string or a transition, I would still prefer using new a built-in function here.

**Reply by katre** [reply]
I initially wanted functional composition, like `composed = trans1(trans2)`, which matches my intuition that a transition is just a function. However, everyone I showed that too agreed it was even more confusing.

We can use [any Starlark binary operator ](https://github.com/bazelbuild/starlark/blob/master/spec.md#binary-operators) with not a lot of extra work.

I'm actually really against adding a new API surface for this, as a general principle. It's already really confusing having both [`config`](https://bazel.build/rules/lib/toplevel/config) and [`config_common`](https://bazel.build/rules/lib/toplevel/config_common). (And, actually, I now think that `config.exec` should have been `config_common.exec`, or that we should just collapse the two together, but API changes are hard).

**Reply by gregestren** [reply]
> However, everyone I showed that too agreed it was even more confusing.

How so? My quick reaction is `composed = trans1(trans2)` looks like it's preemptively applying `trans1` with an input of `trans2`. 

So a builtin alternative would be something like. `composed = config.compose(trans1, trans2)`? I agree `+` is tighter and cleaner, even with its slightly unintuitive semantics. Even more so for multi-composition:

    composed = trans1(trans2(trans3))
    composed. = trans1 + trans2 + trans3

Speaking of, multi-composition should work fine, right? Can we clarify in the doc?

> that we should just collapse the two togethe

+1. `config` + `config_common` seems totally unnecessary.

**Reply by katre** [reply]
The main issue with function application is ordering. `trans1 + trans2 + trans3` makes the order very clear, whereas the equivalent with function application would be `trans1(trans2)(trans3)`, which is fairly tricky to read.

**Reply by gregestren** [reply]
Just to be clear, I agree function application doesn't work well, at least not in `trans1(trans2)` form. Thanks for clarifying.

I support `+`.

**Reply by brandjon** [reply]
Starlark team weighing in: We dislike the use of operator overloading for non-core datatypes (list/tuple/etc). We have to accept this for `select()` for historic reasons, but would prefer not to add any new instances of operator overloading.

The rationale is that it's not always obvious to a user what the type of the arguments is, or even that the usage of the operator is an overloaded one. It's also more difficult to discover the appropriate API docs for an operator through search or browsing API reference pages, compared with an alphanumeric method name.

Making transitions callable (really, overloading the call operator) as in `trans1(trans2)` is better but not great. We'd prefer something like `config.compose(trans1, trans2)` or `trans1.composeWith(trans2)`.

Note that lack of commutativity is not a concern, because that's already the case for lists, strings, etc.

**Reply by katre** [reply]
Updated to call the new function `config.compose_transitions`.

Separately, we should move `config_common` items into `config`.

**Reply by fmeum** [reply]
The updated proposal refers to `config` rather than `config_common`, is that intentional?

A method on `transition` seems more discoverable to me than a global, especially with the advent of Starlark language servers with type support. 

How about (inspired by Java) `transition1.and_then(transition2)`? This also makes the order of application clear (left to right), whereas `compose_with` would probably mean "right to left", but this is less intuitive.

Edit: @katre didn't see your previous comment in time.

**Reply by gregestren** [reply]
Re deprecating `config_common`, I see 90 references in google and 44 references in [bazelbuild](https://github.com/search?q=org%3Abazelbuild+config_common&type=code). Looks manageable.

@fmeum are you suggesting keeping the distinction?

Otherwise I like your suggestion, on both points. @brandjon any general Starlark concerns about that?

**Reply by fmeum** [reply]
No, I'm actually in favor of merging both into `config`, but my understanding of the discussion up to that point was that `config_common` should be kept instead.

**Reply by katre** [reply]
> A method on transition seems more discoverable to me than a global, especially with the advent of Starlark language servers with type support.

This is a valid point, but then keeps the issue with trying to perform a transition starting from the "exec" string.

I am fine with the method on a transition, but then we should entirely drop support for the legacy string names for clarity.

How does everyone feel about this?

**Reply by fmeum** [reply]
@katre Sorry, I accidentally edited your comment on mobile. I restored it.

My reply:

> I am fine with the method on a transition, but then we should entirely drop support for the legacy string names for clarity.

Do you mean even disallowing cfg = "exec" on rules? That would be very breaking - why would that need to happen as a precondition?

**Reply by katre** [reply]
No, those we would leave, but for composition you'd need to use an actual transition object. Is that too restrictive?

**Reply by fmeum** [reply]
> No, those we would leave, but for composition you'd need to use an actual transition object. Is that too restrictive?

No, that sounds good!

**Reply by gregestren** [reply]
Will `config.exec()` seamlessly work in the implementation? That's an `ExcecutionTransitionFactory` and `transition` is a `ConfigurationTransitionApi` which if I read correctly is a distinct interface/class.

**Reply by katre** [reply]
It will require some coding to make it work (and either some overloads or a common superinterface) but it should be supportable.

Or we could get rid of transitions and only use transition factories, but that's a fairly big change IIRC.

I'll clarify the spec on this point as part of marking it accepted.

**Comment by katre** [comment]
Pushed an update to clarify the precise semantics. @fmeum, I know you were interested, let me know if this is more clear.

**Reply by fmeum** [reply]
I think it is, thanks!

**Comment by gregestren** [comment]
LGTM on latest revision.

**Comment by katre** [comment]
More than a year later, an update:

I started trying to implement this last year, and was swamped by the intersection of transition semantics and the underlying Java types used for [`TransitionFactory`](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/analysis/config/transitions/TransitionFactory.java?ss=bazel&q=TransitionFactory).

To recap, the semantics and types involved:
1. An attribute transition can be a patch or split transition, and has access to the target's attribute values (possibly after resolving unrelated configurable attributes via `select`)
   1. This is represented in Java by the `TransitionFactory<AttributeTransitionData>` type
2. A rule transition can only be a patch transition, and has access to the targets raw attribute values (with some partial resolution for configurable attributes but not entirely).
   1. This is represented in Java by the `TransitionFactory<RuleAttributeTransition>` type.
3. A patch transition that does not use target attribute values can be used for either an attribute or a rule.
   1. This is represented in Java by the `TransitionFactory<TransitionFactory.Data>` type.

Except almost all of those statements are false.

**Native** transitions use those types, and can correctly distinguish patch transitions from split transitions. **Starlark** transitions are all instances of [`StarlarkDefinedConfigTransition`](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/analysis/config/StarlarkDefinedConfigTransition.java;drc=f08d561385a5bcbd2f7e5341820fe35cc813bdd3;l=78), which is then wrapped in a `StarlarkAttributeTransitionProvider` or `StarlarkRuleTransitionProvider` when the transition is declared (with `attr.label(cfg = transition)` or `rule(cfg = transition)`, respectively).

Trying to implement a version of `ConfigurationTransitionApi.andThen` or `StarlarkConfig.compose` that handles all of these types and checks the correct semantics became far too complex, and so the work has stalled.

A useful first step to regularize the types involved: trying to express the transition semantics with Java types doesn't work, and so changing everything to use the Starlark API-style distinction between "transition implementation" and "transition usage" is probably worth doing.

I'd like to resume tackling this in the future, but it's fairly unclear when.
